### PR TITLE
Several refactors and optimizations 

### PR DESCRIPTION
**Pull request by Sija**
Proper fix for https://github.com/mint-lang/mint/commit/364af52f3d7bf139f008592db3ffd49859ec0480#r38111985 among others

### CODE REVIEW

File: src/ast/data.cr [hunk 1]
```diff
@@ -6,13 +6,8 @@ module Mint
...
 
-      def to_s
-        "<#{file} #{@input[0, 10]}...>"
-      end
-
-      def to_s(io)
-        io << to_s
-        io
+      def to_s(io : IO)
+        io << '<' << file << ' ' << input[0, 10] << "...>"
```
Comment (gdotdesign) [inline src/ast/data.cr]: Curious why do we need to do this (`<<`) for every part?

Comment (Sija) [reply src/ast/data.cr]: To avoid string interpolations.

**Comment by Sija** [comment]
Sorry for all of the force-pushes, should be GTG now.

**Comment by Sija** [comment]
@gdotdesign Added couple of other optimizations, I hope you don't mind... xD

File: src/ast/data.cr (hunk 1 above)
Comment (Sija) [reply src/ast/data.cr]: Sure! For in-depth explanation see [_Don't create intermediate strings when writing to an IO_](https://crystal-lang.org/reference/guides/performance.html#dont-create-intermediate-strings-when-writing-to-an-io) section in the crystal reference guide.

**Comment by gdotdesign** [comment]
Thank you very much :smile:  :+1:

**Pull request merged**